W&B App: Rewrite _index.md for custom charts (#1685)
wandb/docs: content/en/guides/models/app/features/custom-charts/_index.md

- url: guides/app/features/custom-charts/:filename
---

> Custom charts are visualizations you create explicitly. Use custom charts to visualize more complex data relationships and have greater control over the appearance and behavior of the chart.

Reviewer: I bet you could do without "more". Wdyt?

To create a custom chart, you can either:

> * [Build a chart in the W&B App](#create-a-custom-chart-in-the-wb-app) by pulling in data from your runs using a [GraphQL](https://graphql.org) query.

Reviewer: ```suggestion
* [Build a chart in the W&B App](#create-a-custom-chart-in-the-wb-app) m data from your runs using a [GraphQL](https://graphql.org) query.
```

Reviewer: Oops that was supposed to be "from" not just "m".

* Log a `wandb.Table` and call the `wandb.plot()` function from your script.

{{% alert title="Default charts" %}}
Default charts are generated visualizations based on the data logged from your script with `wandb.Run.log()`. For example, consider the following code snippet:

```python
import wandb
import math

with wandb.init() as run:
  offset = random.random()
  for run_step in range(20):
    run.log({
        "acc": math.log(1 + random.random() + run_step) + offset,
        "val_acc": math.log(1 + random.random() + run_step) + offset * random.random(),
    })
```

> The code snippet logs two metrics `acc` and `val_acc` at each training step (each `wandb.Run.log()` call). In the W&B App, navigate to the Workspace tab. Within the `Charts` panel, two line plots appear: one for `acc` and one for `val_acc`. The x-axis is the `step` value (automatically tracked by W&B), and the y-axis is the metric value you provide with `wandb.Run.log()`.

Reviewer: ```suggestion
This code logs two metrics `acc` and `val_acc` at each training step (each `wandb.Run.log()` call). In the W&B App, navigate to the Workspace tab. Within the `Charts` panel, two line plots appear: one for `acc` and one for `val_acc`. The x-axis is the `step` value (automatically tracked by W&B), and the y-axis is the metric value you provide with `wandb.Run.log()`.
```
More importantly, I think a stepped procedure is buried on this paragraph.

Suppose you want to create a line plot that shows the accuracy as a function of training step. To do this, you first log accuracy values to a run using the following code snippet:

```python
import wandb
import math

with wandb.init() as run:
  offset = random.random()
  for run_step in range(20):
    run.log({
        "acc": math.log(1 + random.random() + run_step) + offset,
        "val_acc": math.log(1 + random.random() + run_step) + offset * random.random(),
    })
```

Each call to `wandb.Run.log()` creates a new entry in the run's history. The `step` value is automatically tracked by W&B and increments with each log call.

Once the code snippet completes, navigate to your project's workspace:

<!-- 
In the W&B App, navigate to the Workspace tab. Within the `Charts` panel, two line plots appear: one for `acc` and one for `val_acc`. The x-axis is the `step` value (automatically tracked by W&B), and the y-axis is the metric value you provide with `wandb.Run.log()`. -->

<!-- Suppose you want to create a line plot of the accuracy values you logged (`acc`) vs step. To do this: -->

> 1. Click on the **Add panels** button in the top right corner, then select **Custom chart**.

Reviewer: ```suggestion
1. Click the **Add panels** button in the top right corner, then select **Custom chart**.
```

2. Select **Line plot** from the list of chart types.
3. Within the [query editor](#query-editor), select `history` as the [data source](#query-data-sources). Next, select `acc` and type in `_step` as keys. 
4. Within the chart editor, select `_step` for the **X** field and `acc` for the **Y** field. Optionally, set a title for the chart. Your settings should look similar to the following:
{{< img src="/images/app_ui/custom-charts-query-example.png" alt="Custom line plot settings" max-width="90%" >}}

> Note that only the keys you specify in the query editor are available in the chart editor (below the query editor). If you do not see the expected keys, go back to the query editor and ensure you have selected the correct data source and specified the correct keys.

Reviewer: Maybe this would be better as an actual note since it might clear up confusion.

* `config`: Initial settings of your experiment (your independent variables). This includes any named fields you’ve logged as keys to `wandb.Run.config` at the start of your training. For example: wandb.Run.config.learning_rate = 0.0001
* `summary`: A single-value metrics that summarize a run. It's populated by the final value of metrics logged with `wandb.Run.log()` or by directly updating the run.summary object. Think of it as a key-value store for your run's final results.
* `history`: Time series data logged with `wandb.Run.log()`. Each call to `log()` creates a new row in the history table. This is ideal for tracking metrics that change over time, like training and validation loss.
* `summaryTable`: A table of summary metrics. It's populated by logging a `wandb.Table` to the `summary` field. This is useful for logging metrics that are best represented in a tabular format, like confusion matrices or classification reports.
* `historyTable`: A table of time series data. It's populated by logging a `wandb.Table` to the `history` field. This is useful for logging complex metrics that change over time, like per-epoch evaluation metrics.

> To summarize, `summary` and `history` are the general locations for your run data, while `summaryTable` and `historyTable` are the specific query types needed to access tabular data stored in those respective locations.

Reviewer: I don't love summarize and summary together

{{% /alert %}}

The `keys` argument acts as a filter to specify exactly which pieces of data you want to retrieve from a larger data object such as `summary`. The sub-fields are dictionary-like or key-value pairs.

The general structure of the query is as follows:

```graphql
query {
    runSets: (runSets: "${runSets}", limit: 500 ) {
        id: 
        name: 
        summary:
            (keys: [""])
        history:
            (keys: [""])
        summaryTable:
            (keys: [""])
        historyTable:
            (keys: [""])
    }
}
```

Here's a breakdown of the components:

* `runSets`: Top-level object, representing the set of runs you are currently viewing or have filtered in the UI.
* `summary(...)`, `history(...)`, `summaryTable(...)`, `historyTable(...)`: This tells the query to fetch data from the respective objects of each run.
* `keys: [""]`: An array of strings where each string is the name (or key) of the metric or object you want to retrieve.

## Log charts from a script

You can programmatically create a custom chart from your script by logging a `wandb.Table` of the data you want to visualize, then calling `wandb.plot.*` to create the chart. 

> For example consider the following code snippet:

Reviewer: ```suggestion
For example, consider the following code snippet:
```

    # Create a table with the columns to plot
    table = wandb.Table(data=data, columns=["step", "height"])

    # Use the table to populate various custom charts
    line_plot = wandb.plot.line(table, x='step', y='height', title='Line Plot')
    histogram = wandb.plot.histogram(table, value='height', title='Histogram')
    scatter = wandb.plot.scatter(table, x='step', y='height', title='Scatter Plot')

    # Log custom tables, which will show up in customizable charts in the UI
    run.log({'line_1': line_plot, 
                'histogram_1': histogram, 
                'scatter_1': scatter})
```

Within the W&B app, navigate to the **Workspace** tab. Within the `Custom Charts` panel, there are three charts with the following titles: **Scatter Plot**, **Histogram**, and **Line Plot**. These correspond to the three charts created in the script above. The x-axis and y-axis are set to the columns specified in the `wandb.plot.*` function calls (`height` and `step`). 

The following image shows the three custom charts created from the script:

{{< img src="/images/app_ui/custom-charts-script-plots.png" alt="Custom charts from script" max-width="90%" >}}

> ### Builtin chart types

Reviewer: Do we tend to use the unhyphenated form? I associate it with bash.
